swiftnav/src/time/mod.rs

```diff
@@ -0,0 +1,181 @@
+//! in to swftnav-rs. This list is kept up to date in the source code as new leap
+//! seconds are announced, but once the code is compiled there is no way to update
+//! this table with new leap seconds. This obviously means that sooner or later
+//! the hard coded leap second information will become out of date and the
+//! converted times will be inaccurate. This is fine if you are processing
+//! historical data, but processing live data runs the risk of an incorrect time
+//! conversion.
+//!
+//! When converting to or from a time base that uses leap seconds (e.g. [`UtcTime`])
+//! two functions are always provided, one which takes a  [`UtcParams`] object to
+//! handle the leap second conversion and one which doesn't take a [`UtcParams`]
+//! object but has `_hardcoded` appended to the function name.
+
+use std::time::Duration;
+
+pub mod consts;
+mod gnss;
+mod mjd;
+mod utc;
+
+pub use gnss::*;
+pub use mjd::*;
+pub use utc::*;
+
+/// A minute long [`Duration`]
+pub const MINUTE: Duration = Duration::from_secs(consts::MINUTE_SECS as u64);
+/// An hour long [`Duration`]
+pub const HOUR: Duration = Duration::from_secs(consts::HOUR_SECS as u64);
+/// A day long [`Duration`]
+pub const DAY: Duration = Duration::from_secs(consts::DAY_SECS as u64);
+/// A week long [`Duration`]
+pub const WEEK: Duration = Duration::from_secs(consts::WEEK_SECS as u64);
+
+/// Checks to see if a year is a leap year
+///
+/// # Note
+///
+/// All year values are treated as if they are in the Gregorian calendar
+pub fn is_leap_year(year: u16) -> bool {
+    ((year % 4 == 0) && (year % 100 != 0)) || (year % 400 == 0)
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    #[test]
+    fn is_leap_year() {
+        use super::is_leap_year;
+        assert!(!is_leap_year(2019));
+        assert!(is_leap_year(2020));
+        assert!(!is_leap_year(1900));
+        assert!(is_leap_year(2000));
+    }
+
+    #[test]
+    fn conversions() {
+        const TEST_CASES: [GpsTime; 10] = [
+            GpsTime::new_unchecked(1234, 567890.0),
+            GpsTime::new_unchecked(1234, 567890.5),
+            GpsTime::new_unchecked(1234, 567890.0),
+            GpsTime::new_unchecked(1234, 0.0),
+            GpsTime::new_unchecked(1000, 604578.0),
+            GpsTime::new_unchecked(1001, 222.222),
+            GpsTime::new_unchecked(1001, 604578.0),
+            GpsTime::new_unchecked(1939, 222.222),
+            GpsTime::new_unchecked(1930, 16.0),
+            GpsTime::new_unchecked(1930, 18.0), /* around Jan 2017 leap second */
+        ];
+
+        const TOW_TOL: f64 = 1e-6;
+
+        for test_case in TEST_CASES {
+            // test gps -> mjd -> gps
+            let mjd = test_case.to_mjd_hardcoded();
+            let round_trip = mjd.to_gps_hardcoded();
+            let diff = test_case.diff(&round_trip).abs();
+            assert!(
+                diff < TOW_TOL,
+                "gps2mjd2gps failure. original: {:?}, round trip: {:?}, diff: {}, TOW_TOL: {}",
```

Comment:
Can we use rust's face new inline feature for these asserts?




```suggestion
assert!(
   diff < TOW_TOL,
   "gps2mjd2gps failure. original: {test_case:?}, round trip: {round_trip:?}, diff: {diff}, TOW_TOL: {TOW_TOL}"
);
```

Reply:
That's a feature new in 1.89, right?

Reply:
Oh yeah, just realized that this crate is on edition 2018. I'm going to leave this as is for the moment, and when we bump the edition (which we should) I think these should get caught by a clippy lint (especially if we enable pedantic)